File: 606_construct-string-from-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    string tree2str(TreeNode* t) {
        if (!t) return "";
        string left = tree2str(t->left);
        string right = tree2str(t->right);
        stringstream ss;
        ss << t->val;
        if (left.empty() && right.empty()) return ss.str();
        if (right.empty()) {
            ss << '(' << left << ')';
            return ss.str();
        }
        ss << '(' << left << ')' << '(' << right << ')';
        return ss.str();
    }
};

File: 606_construct-string-from-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """
        if not t:
            return ''
        left = self.tree2str(t.left)
        right = self.tree2str(t.right)
        if len(right):
            return ''.join([str(t.val), '(', left, ')', '(', right, ')'])
        elif len(left):
            return ''.join([str(t.val), '(', left, ')'])
        else:
            return str(t.val)

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """
        if not t: return ''
        left = '({})'.format(self.tree2str(t.left)) if (t.left or t.right) else ''
        right = '({})'.format(self.tree2str(t.right)) if (t.right) else ''
        return '{}{}{}'.format(t.val, left, right)
